spec/unit/deep_preloader_spec.rb

```diff
@@ -0,0 +1,259 @@
+        DeepPreloader.preload(parent, :child)
+        expect(parent).to have_loaded(:child).as(child)
+      end
+
+      it "sets up the inverse relationship" do
+        DeepPreloader.preload(parent, :child)
+        expect(parent.child).to have_loaded(:parent).as(parent)
+      end
+
+      it "satisfies a childless parent" do
+        DeepPreloader.preload(childless, :child)
+        expect(childless).to have_loaded(:child).as(nil)
+      end
+
+      it "loads more than one entity" do
+        DeepPreloader.preload([parent, childless], :child)
+        [parent, childless].zip([child, nil]).each do |p, c|
+          expect(p).to have_loaded(:child).as(c)
+        end
+      end
+
+      it "uses preloaded children from other parents before hitting the database" do
+        loaded_parent = Parent.create!(child: child)
+        Child.where(id: child).delete_all
+
+        DeepPreloader.preload([parent, loaded_parent], :child)
+        [parent, loaded_parent].each do |p|
+          expect(p).to have_loaded(:child).as(child)
+        end
+      end
+
+      it "supports locking" do
+        DeepPreloader.preload(parent, :child, lock: "FOR SHARE")
+        # No real way to test this with sqlite: AR drops the lock clause.
+      end
+    end
+
+    context "in the has_one direction" do
+      let!(:parentless) { Child.create! }
+
+      it "loads the parent" do
+        DeepPreloader.preload(child, :parent)
+        expect(child).to have_loaded(:parent).as(parent)
+      end
+
+      it "sets up the inverse relationship" do
+        DeepPreloader.preload(child, :parent)
+        expect(child.parent).to have_loaded(:child).as(child)
+      end
+
+      it "satisfies a parentless child" do
+        DeepPreloader.preload(parentless, :parent)
+        expect(parentless).to have_loaded(:parent).as(nil)
+      end
+
+      it "loads more than one entity" do
+        DeepPreloader.preload([child, parentless], :parent)
+        [child, parentless].zip([parent, nil]).each do |c, p|
+          expect(c).to have_loaded(:parent).as(p)
+        end
+      end
+
+      it "handles a preloaded child" do
+        child.parent # force load
+        DeepPreloader.preload(child, :parent)
+        expect(child).to have_loaded(:parent).as(parent)
+      end
+    end
+  end
+
+  context "with multiple belongs_to relationships" do
+    with_temporary_table(:child) do
+    end
+
+    with_temporary_table(:parent1, ->(t){ t.references :a_child; t.references :b_child }) do
+      belongs_to :a_child, class_name: Child.name
+      belongs_to :b_child, class_name: Child.name
+    end
+
+    with_temporary_table(:parent2, ->(t){ t.references :child }) do
+      belongs_to :child
+    end
+
+    let!(:c1) { Child.create! }
+    let!(:c2) { Child.create! }
+    let!(:c3) { Child.create! }
+    let!(:p1) { Parent1.create!(a_child: c1, b_child: c2).tap(&:reload) }
+    let!(:p2) { Parent2.create!(child: c3).tap(&:reload) }
+    before(:each) { [c1, c2, c3].map(&:reload) }
```

Comment:
When you create a parent and child together, the association between parent and child is preloaded. We don't want that, so we have to `reload`.